Engage pages from discourse

carkod opened the pull request:
## Done

- Extended engage pages to use discourse
- Index page uses now discourse
- Individual engage pages have webinar image pull from brighttalk
- Added thank-you pages for the forms and related engage pages.

## QA

- Go to https://ubuntu-com-8086.demos.haus/engage and click the different engage pages
- CTA should not show for large screens, if 2 CTAs should show on large screens. Otherwise it should only display on small screens.
- CTA buttons should link to respective elements on the page
- Fill in some forms and see that they take you to the thank-you page.
- If you have a Brittalk account, login to check that the video plays correctly.

## Issues

Fix https://github.com/canonical-web-and-design/web-squad/issues/3035
Fix https://github.com/canonical-web-and-design/web-squad/issues/3034

webteam-app commented:
Demo starting at https://ubuntu-com-8086.demos.haus

nottrobin commented:
@carkod you could make this easier to QA by updating `site.yaml` to read the environment variable from a secret, and then set a secret in the demo cluster to give it access to discourse. I'd recommend you use an API key for your own user maybe.

carkod commented:
> @carkod you could make this easier to QA by updating `site.yaml` to read the environment variable from a secret, and then set a secret in the demo cluster to give it access to discourse. I'd recommend you use an API key for your own user maybe.

I don't think the demo is running, because we are using a workaround to install a Git branch as a `requirements.txt` dependency so must be QAed locally

carkod commented:
@steverydz since you are the master of takeovers and engage pages, can you take a look at some of the following points in this PR?:
- In the individual engage pages /engage/<engage-page-name> the `p-strip` styling works great for some engage pages but not so well for some others, e.g. colored or gradient backgrounds. Not sure if there is a vanilla modifier class for this, if not I guess we need to add it to vanilla?
- Ubuntu logos for individual engage pages. We don't have data in [discourse](https://discourse.ubuntu.com/t/about-the-engage-pages-category/17229) to populate these, so I thought maybe we have some standard white and black logos, likes the one we currently use, but I don't know they are stable enough to use them for this purpose (they look generated?).

steverydz commented:
> * In the individual engage pages /engage/ the `p-strip` styling works great for some engage pages but not so well for some others, e.g. colored or gradient backgrounds. Not sure if there is a vanilla modifier class for this, if not I guess we need to add it to vanilla? e.g. engage/cloud-init-whitepaper

It looks like the engage page headers aren't using the right class names, e.g. http://localhost:8001/engage/fr/migrer-de-vmware-a-openstack is using `p-strip--dark` when it should be `p-strip p-engage-banner--dark`.

Another example is http://localhost:8001/engage/intro-to-anbox-cloud-whitepaper using `p-strip--grad` when it should be `p-strip p-engage-banner--grad`. I'm not sure how you're getting the classes that are used here but it seems the pattern should be replacing `p-strip--dark` etc with `p-strip p-engage-banner--dark` and so on.

> * Ubuntu logos for individual engage pages. We don't have data in [discourse](https://discourse.ubuntu.com/t/about-the-engage-pages-category/17229) to populate these, so I thought maybe we have some standard white and black logos, likes the one we currently use, but I don't know they are stable enough to use them for this purpose (they look generated?).

The logos used are just from the asset server so should be fine to carry on using

carkod commented:
> > In the individual engage pages /engage/ the `p-strip` styling works great for some engage pages but not so well for some others, e.g. colored or gradient backgrounds. Not sure if there is a vanilla modifier class for this, if not I guess we need to add it to vanilla? e.g. engage/cloud-init-whitepaper
> 
> It looks like the engage page headers aren't using the right class names, e.g. http://localhost:8001/engage/fr/migrer-de-vmware-a-openstack is using `p-strip--dark` when it should be `p-strip p-engage-banner--dark`.
> 
> Another example is http://localhost:8001/engage/intro-to-anbox-cloud-whitepaper using `p-strip--grad` when it should be `p-strip p-engage-banner--grad`. I'm not sure how you're getting the classes that are used here but it seems the pattern should be replacing `p-strip--dark` etc with `p-strip p-engage-banner--dark` and so on.
> 
> > Ubuntu logos for individual engage pages. We don't have data in [discourse](https://discourse.ubuntu.com/t/about-the-engage-pages-category/17229) to populate these, so I thought maybe we have some standard white and black logos, likes the one we currently use, but I don't know they are stable enough to use them for this purpose (they look generated?).
> 
> The logos used are just from the asset server so should be fine to carry on using

Thanks. Can you take a look at the thank you pages too (navigate to it after submitting a form)?

- Do we need a resource name or is it enough to just write the type? (e.g. whitepaper)
- There are some "related engage pages" below, how many do you think we should keep? Current page.ubuntu.com only has 3 but we can show more

steverydz commented:
@carkod Unless there has been a request for a design change I'd recommend using the same classes as the live thank you page.

> Do we need a resource name or is it enough to just write the type? (e.g. whitepaper)
I'm guessing type will be what we want?

> There are some "related engage pages" below, how many do you think we should keep? Current page.ubuntu.com only has 3 but we can show more
I think we should stick with the same format as live unless someone has requested that there should be more

anthonydillon commented on templates/engage/thank-you.html:
```diff
@@ -0,0 +1,76 @@
+{% extends "engage/base_engage.html" %}
+
+{% block title %}Engage download "resource_name"{% endblock %}
```
"resource_name"?

carkod replied:
Fixed

anthonydillon commented on templates/engage/thank-you.html:
```diff
@@ -0,0 +1,76 @@
+  </div>
+  <div class="row">
+    <div class="col-8">
+      <div>
```
Can remove this div

anthonydillon commented on templates/engage/thank-you.html:
```diff
@@ -0,0 +1,76 @@
+  </div>
+</section>
+
+<section class="p-string--light">
```
`strip`

anthonydillon commented on templates/engage/thank-you.html:
```diff
@@ -0,0 +1,76 @@
+<section class="p-string--light">
+  <div class="row">
+    <div class="col-8">
+      <h2 class="p-heading--three">You may also be interested in …</h2>
```
Please encode `...` with `&hellip;`

anthonydillon commented on templates/engage/thank-you.html:
```diff
@@ -0,0 +1,76 @@
+      <!-- THREE ADDITIONAL CTAs -->
+      <h4>{{page["topic_name"]}}</h4>
+      <p>{{page["subtitle"]}}</p>
+      <p><a href="{{page['path']}}">See more›</a></p>
```
```suggestion
      <p><a href="{{page['path']}}">See more&nbsp;&rsaquo;</a></p>
```

anthonydillon reviewed (changes requested):
Some issues in the codebase. Need a full QA.

carkod commented on templates/engage/base.html:
```diff
@@ -0,0 +1,109 @@
+{% set section_title="Engage Page" %}
+{% set section_path="/engage" %}
+
+{% block title %}ESM for Ubuntu 14.04 LTS Trusty Tahr{% endblock %}
```
Fixed

carkod commented on templates/engage/base.html:
```diff
@@ -0,0 +1,109 @@
+{% set section_path="/engage" %}
+
+{% block title %}ESM for Ubuntu 14.04 LTS Trusty Tahr{% endblock %}
+{% block meta_description %}Extended Security Maintenance (ESM) will be available for Ubuntu 14.04 LTS Trusty Tahr for continued security and compliance of the most popular linux operating system.{% endblock %}
```
We don't have meta description data. I am using now the subtitle by default, but not all engage pages have it.

carkod commented on templates/engage/base.html:
```diff
@@ -0,0 +1,109 @@
+      {% if document["metadata"]["webinar_code"] != "" %}
+      <p>
+        <a href="#register-section" class="p-button--positive">
+          Watch the webinar
```
- We don't have CTA text. I added that text as default now, and check if there is CTA if not, output that default text.
- Regarding link, in this specific case there is no need. It's an anchor that takes you to the brighttalk webinar section that generates the link, like in our current engage pages

carkod replied:
So I thought of maybe a better way to do at least the download CTA button text. It seems to be always based on the type property in the discourse data. So it's always something like Download the {type}, so you will get eventually `Download the whitepaper` or `Download the ebook`. For things like events or webinar, there is no formid, so it will not display the form.

The other CTA would be "Watch the webinar" which might need translation...

carkod replied:
I rectify myself, we still have the translation problem...

carkod commented on templates/engage/base.html:
```diff
@@ -0,0 +1,109 @@
+    <div class="col-4 u-hide--small u-vertically-center u-align--center">
+      
+      {{
+        image(
```
Done. An issue is that the image can be too big and take up too much width. Maybe keep the `width` property? or is too much width not a problem?

carkod commented on templates/engage/base.html:
```diff
@@ -0,0 +1,109 @@
+    {% if document["metadata"]["form_id"] != "" %}
+    <div class="col-5" id="the-form">
+      {% if document["metadata"]["language"] != '' %}
+        {% with id=document["metadata"]["form_id"], returnURL="/engage/thank-you", cta="Download the case study" %}
```
I suppose that you mean is the CTA? everything else comes from the data source. In which case, it's done.

carkod commented:
> Some issues in the codebase. Need a full QA.

Addressed most of the comments. The ones left are data-related. Probably the important one to consider is the CTA

jpmartinspt commented on konf/site.yaml:
```diff
@@ -36,6 +36,16 @@ env:
 
+  - name: DISCOURSE_API_KEY
+    secretKeyRef:
+      key: api_key
```
can we swap to `api-key`?

jpmartinspt commented on konf/site.yaml:
```diff
@@ -36,6 +36,16 @@ env:
+
+  - name: DISCOURSE_API_USERNAME
+    secretKeyRef:
+      key:  api_user
```
can we swap to `api-user`?

jpmartinspt commented on konf/site.yaml:
```diff
@@ -36,6 +36,16 @@ env:
+  - name: DISCOURSE_API_KEY
+    secretKeyRef:
+      key: api_key
+      name: discourse-api
```
The secret is named `discourse-api` for demos, but for production and staging you went with just `discourse`.. It needs to be the same everywhere and updated here if you choose to go with `discourse`.

carkod commented:
@anthonydillon can you take another look when you can? I added a few more paramters e.g. `primary_cta`, `primary_link` and also `form_cta` and `channel_id` to fix stuff in pages with different languages. You reminded me yesterday of the thank-you pages which I have added in all our currently supported languages.

I reviewed all foreign language engage pages and about 20 other English pages and haven't seen major issues.

Do not merge as I want to double check secrets before we deploy, and if everything is fine I will add these new parameters in the discourse guide.

anthonydillon commented:
## QA
After a quick look around this engage page has a few issues: https://ubuntu-com-8086.demos.haus/engage/finserv-multi-cloud
- The first sentence has a line break-in it.
- The form leads to a 500

carkod commented:
> ## QA
> After a quick look around this engage page has a few issues: https://ubuntu-com-8086.demos.haus/engage/finserv-multi-cloud
> 
> * The first sentence has a line break-in it.
> * The form leads to a 500

Fixed.

anthonydillon commented on templates/engage/thank-you.html:
```diff
@@ -0,0 +1,75 @@
+  </div>
+</section>
+
+{% if related | length > 0%}
```
Space after the `0`

anthonydillon reviewed (changes requested):
QA issue:
- Downloading this engage page https://ubuntu-com-8086.demos.haus/engage/nfv-sdn-openstack gets me the MAAS ebook not the one from the resource_url in [the post](https://discourse.ubuntu.com/t/nfv-and-sdn-on-openstack-for-network-operators/17326).

carkod commented:
> QA issue:
> 
> * Downloading this engage page https://ubuntu-com-8086.demos.haus/engage/nfv-sdn-openstack gets me the MAAS ebook not the one from the resource_url in [the post](https://discourse.ubuntu.com/t/nfv-and-sdn-on-openstack-for-network-operators/17326).

The MAAS ebook? I can't reproduce. I keep getting the NFV ebook

anthonydillon commented:
Maybe the demo is cached which will be a problem in prod.

anthonydillon commented:
I think you will need to change the thank you system to have unique URLs. For example, /engage/{engage_slug}/thank-you and just include the thank you template.

jpmartinspt commented on webapp/views.py:
```diff
@@ -674,6 +674,71 @@ def tutorials_index():
+
+    def render_template():
+
+        if hasattr(engage_pages.parser, "current_topic"):
```
I'm not entirely convinced we should be rendering a template based on the state of a property of a `global` object.
Can we just  pass the `topic_id` as a parameter to the thank you view and get the info in the view?

carkod replied:
That will change with the new way we are rendering the thank-you page. I will update it.

jpmartinspt commented on webapp/views.py:
```diff
@@ -674,6 +674,71 @@ def tutorials_index():
     return tutorials_index
 
 
+def build_engage_index(engage_docs):
```
Is there a reason this view can't be part of the discourse module itself?

carkod replied:
so in a discussion we had about how to structure the module, we decided that what is essentially project stuff would be handled by the project and the module will only handle the data. This is also how it's done on tutorials if you look at above that function currently

carkod commented:
> QA issue:
> 
> * Downloading this engage page https://ubuntu-com-8086.demos.haus/engage/nfv-sdn-openstack gets me the MAAS ebook not the one from the resource_url in [the post](https://discourse.ubuntu.com/t/nfv-and-sdn-on-openstack-for-network-operators/17326).

Fixed

anthonydillon commented on templates/engage/base.html:
```diff
@@ -0,0 +1,134 @@
+    </div>
+    {% if document["metadata"]["image"] != '' %}
+    <div class="col-5 u-hide--small u-vertically-center u-align--center">
+      <img src="{{document['metadata']['image']}}" alt="{{document['metadata']['topic_name']}}" width="{{document['metadata']['image_width']}}"/>
```
An empty alt is better then one that has nothing to do with the image

anthonydillon reviewed (commented):
Code looks good with one comment. QA of existing pages look good.

I want to QA some test engage pages to see if I can make the site fall over. Also, check form submitting are going through.

anthonydillon commented:
## QA
- The final few have gone full width: https://ubuntu-com-8086.demos.haus/engage

caldav reviewed (approved):
Looks GREAT overall, one small comment: on engage pages, header buttons should be `u-hide-large` unless there is more than 1. 

Rationale behind this design is that coming from a takeover, the header with a button would look too similar to the takeover and a page change would not necessarily be obvious. We nevertheless want it visible on mobile, as it's used as a fast track to the content. 

When there are two CTAs, it's usually because there are two pieces of content, and we want to make this obvious on desktop.

carkod commented:
> ## QA
> * The final few have gone full width: https://ubuntu-com-8086.demos.haus/engage and dont have links.

Sorry, that was a rebase mess, clicked the vscode button but didn't check if it did it correctly.

carkod commented:
> Looks GREAT overall, one small comment: on engage pages, header buttons should be `u-hide-large` unless there is more than 1.
> 
> Rationale behind this design is that coming from a takeover, the header with a button would look too similar to the takeover and a page change would not necessarily be obvious. We nevertheless want it visible on mobile, as it's used as a fast track to the content.
> 
> When there are two CTAs, it's usually because there are two pieces of content, and we want to make this obvious on desktop.

Thanks. It's done. 
This can be checked viewing the https://ubuntu-com-8086.demos.haus/engage/kubernetes-deployment-enterprise-whitepaper, which will show 2 buttons in large too, but all the others will never display a button in the large screen.

carkod commented:
> Code looks good with one comment. QA of existing pages look good.
> 
> I want to QA some test engage pages to see if I can make the site fall over. Also, check form submitting are going through.

Well, if you mess up the Markdown on the index topic, we cannot show anything, so you can try that and see that the site is still up, but /engage will return 500
Alternatively, you can use my [test index topic](https://discourse.ubuntu.com/t/test-engage-pages/18033) which has the index topic messed up. Just need to change the index topic id.

anthonydillon commented:
@carkod thanks for all this work! Working and looking good. Just need one final sweep to add new engage pages and this should be good to go :+1:
